src/ScaledPolynomialTransformFitter.cc

```diff
@@ -0,0 +1,465 @@
+        detail::computePowers(_transform._poly._u, input.getX());
+        detail::computePowers(_transform._poly._v, input.getY());
+        // We pack coefficients in the following order:
+        // (0,0), (0,1), (1,0), (0,2), (1,1), (2,0)
+        // Note that this lets us choose the just first N(N+1)/2 columns to
+        // evaluate an Nth order polynomial, even if N < maxOrder.
+        for (int n = 0, j = 0; n <= maxOrder; ++n) {
+            for (int p = 0, q = n; p <= n; ++p, --q, ++j) {
+                _vandermonde(i, j) = _transform._poly._u[p] * _transform._poly._v[q];
+            }
+        }
+    }
+}
+
+void ScaledPolynomialTransformFitter::fit(int order) {
+    if (order < 0) {
+        order = _transform.getPoly().getOrder();
+    }
+    int const packedSize = detail::computePackedSize(order);
+    std::size_t nGood = 0;
+    if (_keys.valid.isValid()) {
+        for (auto const & record : _data) {
+            if (record.get(_keys.valid)) {
+                ++nGood;
+            }
+        }
+    } else {
+        nGood = _data.size();
+    }
+    // One block of the block-diagonal (2x2) unweighted design matrix M;
+    // m[i,j] = u_i^{p(j)} v_i^{q(j)}.  The two nonzero blocks are the same,
+    // because we're using the same polynomial basis for x and y.
+    Eigen::MatrixXd m = Eigen::MatrixXd::Zero(nGood, packedSize);
+    // vx, vy: (2x1) blocks of the unweighted data vector v
+    Eigen::VectorXd vx = Eigen::VectorXd::Zero(nGood);
+    Eigen::VectorXd vy = Eigen::VectorXd::Zero(nGood);
+    // sxx, syy, sxy: (2x2) blocks of the covariance matrix S, each of which is
+    // individually diagonal.
+    Eigen::ArrayXd sxx(nGood);
+    Eigen::ArrayXd syy(nGood);
+    Eigen::ArrayXd sxy(nGood);
+    Eigen::Matrix2d outS = _outputScaling.getLinear().getMatrix();
+    for (std::size_t i1 = 0, i2 = 0; i1 < _data.size(); ++i1) {
+        if (!_keys.valid.isValid() || _data[i1].get(_keys.valid)) {
+            afw::geom::Point2D output = _outputScaling(_data[i1].get(_keys.output));
+            vx[i2] = output.getX();
+            vy[i2] = output.getY();
+            m.row(i2) = _vandermonde.row(i1).head(packedSize);
+            if (_keys.outputErr.isValid()) {
+                Eigen::Matrix2d modelErr = outS*_data[i1].get(_keys.outputErr).cast<double>()*outS.adjoint();
+                sxx[i2] = modelErr(0, 0);
+                sxy[i2] = modelErr(0, 1);
+                syy[i2] = modelErr(1, 1);
+            } else {
+                sxx[i2] = 1.0;
+                sxy[i2] = 0.0;
+                syy[i2] = 1.0;
+            }
+            ++i2;
+        }
+    }
+    // Do a blockwise inverse of S.  Note that the result F is still symmetric
+    Eigen::ArrayXd fxx = 1.0/(sxx - sxy.square()/syy);
+    Eigen::ArrayXd fyy = 1.0/(syy - sxy.square()/sxx);
+    Eigen::ArrayXd fxy = -(sxy/sxx)*fyy;
+#ifdef LSST_ScaledPolynomialTransformFitter_TEST_IN_PLACE
+    assert((sxx*fxx + sxy*fxy).isApproxToConstant(1.0));
```

Comment:
I suspect we will disagree about this, but this section is presently on by default and I'm very nervous about asserts being run in production code. I'd be much happier with raising an exception so the user gets a more useful message and other code has a chance to continue if the fitter fails.